File: Java/LeetCode/Matrix Cells in Distance Order.java
// https://leetcode.com/problems/matrix-cells-in-distance-order

class Solution {

    public int[][] allCellsDistOrder(int rows, int cols, int rCenter, int cCenter) {
        LinkedList<LinkedList<Integer>> ans = new LinkedList<>();

        for (int r1 = 0; r1 < rows; ++r1) {
            for (int c1 = 0; c1 < cols; ++c1) {
                ans.add(new LinkedList<>(
                    Arrays.asList(r1, c1, Math.abs(r1 - rCenter) + Math.abs(c1 - cCenter))
                ));
            }
        }

        Collections.sort(ans, new Comparator<LinkedList<Integer>>() {
            @Override
            public int compare(LinkedList<Integer> t1, LinkedList<Integer> t2) {
                return t1.get(2).compareTo(t2.get(2));
            }
        });

        for (LinkedList<Integer> t : ans) {
            t.removeLast();
        }

        int[][] arr = ans.stream()
            .map(list -> list.stream().mapToInt(Integer::intValue).toArray())
            .toArray(int[][]::new);

        return arr;
    }

}

File: Kotlin/LeetCode/Matrix Cells in Distance Order.kt
// https://leetcode.com/problems/matrix-cells-in-distance-order

class Solution {
  
    fun allCellsDistOrder(rows: Int, cols: Int, rCenter: Int, cCenter: Int): Array<IntArray> {
        val ans: MutableList<MutableList<Int>> = mutableListOf()

        for (r1 in 0 until rows) {
            for (c1 in 0 until cols) {
                ans.add(mutableListOf(r1, c1, Math.abs(r1 - rCenter) + Math.abs(c1 - cCenter)))
            }
        }

        ans.sortWith(Comparator { t1, t2 ->
            t1[2].compareTo(t2[2])
        })

        ans.forEach {
            it.removeAt(it.size - 1)
        }

        return ans.map { it.toIntArray() }.toTypedArray()
    }
    
}
